internal/utils/kubernetes/metadata.go

```diff
@@ -15,3 +16,39 @@ import (
 func GetAPIVersionForObject(obj client.Object) string {
 	return fmt.Sprintf("%s/%s", obj.GetObjectKind().GroupVersionKind().Group, obj.GetObjectKind().GroupVersionKind().Version)
 }
+
+// EnsureFinalizerInMetadata ensures the expected finalizer exist in ObjectMeta.
+// If the finalizer does not exist, append it to finalizers.
+// Returns true if the ObjectMeta has been changed.
+func EnsureFinalizerInMetadata(metadata *metav1.ObjectMeta, finalizer string) bool {
+	for _, f := range metadata.Finalizers {
+		if f == finalizer {
+			return false
+		}
+	}
+
+	metadata.Finalizers = append(metadata.Finalizers, finalizer)
+	return true
+}
+
+// RemoveFinalizerInMetadata removes the finalizer from the finalizers in ObjectMeta.
+// If it exists, remove the finalizer from the slice.
+// Returns true if the ObjectMeta has been changed.
+func RemoveFinalizerInMetadata(metadata *metav1.ObjectMeta, finalizer string) bool {
```

Comment:
If the newly introduced type is to have a purpose, it should be accepted in here (the same above in `EnsureFinalizerInMetadata`

```suggestion
// RemoveFinalizerInMetadata removes the finalizer from the finalizers in ObjectMeta.
// If it exists, remove the finalizer from the slice.
// Returns true if the ObjectMeta has been changed.
func RemoveFinalizerInMetadata(metadata *metav1.ObjectMeta, finalizer ControlPlaneFinalizer) bool {
```

WDYT @randmonkey ?

Reply:
The purpose of `RemoveFinalizerInMetadata` is a general function for adding or removing finalizers, not limited to `ControlPlane`. So here I did not use `ControlPlaneFinalizer` here.

Reply:
So perhaps the type shouldn't be so narrow i.e. defined only for controlplane finalizers but generally for finalizers?

This way the funcs you've added (e.g. the one above) could use it and also in the future if there's new finalizers added for other type like e.g. `DataPlane` then it can be reused.

WDYT @randmonkey ?

Reply:
I defined the type `ControlPlaneFinalizer` just to annotate that these are finalizers added to `ControlPlane`s by controlplane controller, but not define a type for all finalizers. If you really, really, really ... concern about the type being defined but not used, I think I could remove the definition of `ControlPlaneFinalizer`, and use string directly when call these functions in controlplane controller.

Reply:
Feel free to resolve this comment if you disagree with this approach.

I'll let others express their view on this.